controller/lib/hal/hal_stm32.cpp

```diff
@@ -292,6 +287,60 @@ void HalApi::delay(Duration d) {
 
 Time HalApi::now() { return millisSinceStartup(msCount); }
 
+/******************************************************************
+ * Loop timer
+ *
+ * I use one of the basic timers (timer 7) to generate the interrupt
+ * from which the high priority loop function is called.
+ *
+ * The basic timers (like timer 7) are documented in chapter 29 of
+ * the reference manual
+ *****************************************************************/
+static void (*high_priority_callback)(void *);
+static void *high_priority_arg;
+void HalApi::startLoopTimer(const Duration &period, void (*callback)(void *),
+                            void *arg) {
+  high_priority_callback = callback;
+  high_priority_arg = arg;
+
+  // Find the loop period in clock cycles
+  int32_t reload = static_cast<int32_t>(CPU_FREQ * period.seconds());
+  int prescale = 1;
+
+  // Adjust the prescaler so that my reload count will fit in the 16-bit
+  // timer.
+  if (reload > 65536) {
+    prescale = static_cast<int>(reload / 65536.0) + 1;
+    reload /= prescale;
+  }
+
+  // Enable the clock to the timer
+  EnableClock(TIMER15_BASE);
+
+  // Just set the timer up to count every microsecond.
+  TimerRegs *tmr = TIMER15_BASE;
+  tmr->reload = reload - 1;
+  tmr->prescale = prescale - 1;
+  tmr->event = 1;
+  tmr->ctrl[0] = 1;
+  tmr->intEna = 1;
+
+  // Enable the interrupt that will call the high priority
+  // callback function periodically.
+  // I'm using a lower priority then that which I use
+  // for normal hardware interrupts.  This means that other
+  // interrupts can be serviced while my high priority
+  // loop functions are running.
+  EnableInterrupt(INT_VEC_TIMER15, IntPriority::LOW);
```

Comment:
This looks very confusing because the comment says "high priority" but the code says Priority::LOW. Is the callback high priority, but not so high as to prevent hardware interrupts? There should be a better naming scheme - perhaps SOFTWARE_HIGH / SOFTWARE_LOW and HARDWARE or something like that? Or even IntPriority::CONTROLLER_LOOP, IntPriority::COMMS etc?